### ISSUE TITLE
Normalize Markdown in `.pot` files

### ISSUE THREAD

**Issue by mgeisler** [google/mdbook-i18n-helpers#19]
When `mdbook-xgettext` extracts translatable text, it would be great if it could normalize the strings. This would make it possible for us to reformat the entire course without fearing that the translations get destroyed while doing so.

The normalization would take Markdown like this

~~~markdown
# This is a heading

This is another heading
=======================

A _little_
paragraph.

```rust,editable
fn main() {
    println!("Hello world!");
}
```

* First
* Second
~~~

and turn it into these messages in the `.pot` file:

* `"This is a heading"` ([atx heading](https://spec.commonmark.org/0.30/#atx-headings) is stripped)
* `"This is another heading"` ([setext heading](https://spec.commonmark.org/0.30/#setext-headings) is stripped)
* `"A _little_ paragraph."` (soft-wrapped lines are unfolded)
* `"fn main() {\n    println!("Hello world!");\n}"` (info string is stripped, we should instead use a [`#,` flag](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html))
* `"First"` (bullet point extracted individually)
* `"Second"`

Like in google/comprehensive-rust#318, we should do this in a step-by-step fashion and make sure to apply the transformations to the existing translations. It would also be good if we have a way to let translators update their not-yet-submitted translations.

**Comment by djmitche** [comment]
I see the appeal, but I'm also wary of making it too easy to add too much code to a slide.  I think a wide-open `.rs` file in a text editor invites adding extra lines.

**Comment by mgeisler** [comment]
> I don't think I'm going to get a chance to work on this issue soon.

That's alright, I'll look at it in the background and see what I can come up with.

**Comment by mgeisler** [comment]
> @mgeisler I'm even hesitant, do you think I should wait a bit with [rust-unofficial/patterns#359](https://github.com/rust-unofficial/patterns/pull/359) and reinitialize the `de.po`-file when this feature is implemented?

I would not wait for this feature. My plan is to make it easy to move to a new version and keep the existing translations will continue to work. People (like yourself!) have put a lot of work into translations and we need to preserve this.

Concretely, I'm planning on providing a small program which people can run on the existing translations to normalize the Markdown found in them. So if the `messages.pot` file or a `xx.po` file contains

```po
#: src/welcome.md:10
msgid ""
"* Give you a comprehensive understanding of the Rust syntax and language.\n"
"* Enable you to modify existing programs and write new programs in Rust.\n"
"* Show you common Rust idioms."
msgstr ""
```

Then the tool will turn the `msgid` field into three, one for each bullet point:

```po
#: src/welcome.md:10
msgid "Give you a comprehensive understanding of the Rust syntax and language."
msgstr ""

#: src/welcome.md:11
msgid "Enable you to modify existing programs and write new programs in Rust."
msgstr ""

#: src/welcome.md:12
msgid "Show you common Rust idioms."
msgstr ""
```

The `msgstr` field will be split the same way, but in this case it was empty. You should upgrade to a new version of mdbook-i18n-helpers and run this program on the existing translations in a single step. Translators can run the same program on their in-progress work.

> it feels a bit inefficient and super fiddly to just move two words in a sentence in the right place (due to \r \n), because the rest is somewhat fine.

I'm not sure why you have to fiddle with individual words like this? Markdown doesn't care about single newlines, so you can break your paragraphs any way you like in the `msgstr` fields. So your `de.po` file can look like this if you want:

```po
msgid "Hallo, I am a little text."
msgstr ""
"Hallo,\n"
"ich bin\n"
"ein kleiner\n"
"Text."
```

The this will result in intermetiary Markdown looking like

```markdown
Hallo,
ich bin
ein kleiner
Text.
```

and this in turn renders into HTML exactly the same way as if it the Markdown had been

```markdown
Hallo, ich bin ein kleiner Text.
```

A different way to put this is that the translators are building up a full Markdown document, but doing it paragraph by paragraph. This implies that the translators must obey the Markdown formatting rules. If you mis-translate the above to

```po
msgid "Hallo, I am a little text."
msgstr ""
"Hallo,\n\n"
"ich bin\n"
"ein kleiner\n"
"Text."
```

Then `mdbook` sees this Markdown

```markdown
Hallo,

ich bin
ein kleiner
Text.
```

Now you have two paragraphs in the final book because of how Markdown uses empty lines to separate paragraphs:

```html
<p>Hallo,</p>

<p>ich bin
ein kleiner
Text.</p>
```

Does that help to explain things?

**Comment by mgeisler** [comment]
> It wouldn't factor in the second line of the translation as it only translates line by line, it seems?

Aha... thanks for explaining, now I understand what you mean!

I don't actually know how or if the line break influences the translations done with `cloud-translate`... in any case, I think removing the newlines would be something best left for that project.

> So my assumption is, that if I would restart and regenerate the translation process, and send everything again via cloud-translate after formatting it would translate it better than now, ending up in less work overall.

From speaking to translators of the Rust course, I think people have had very mixed experiences with using `cloud-translate`: it gets a lot of things wrong because of the specialized context. Perhaps it works better for larger books, I'm not sure.

Feel free to open an issue about this over in https://github.com/mgeisler/cloud-translate — we could perhaps have the tool strip out things like `\n` and other formatting characters.

**Comment by mgeisler** [comment]
I wanted to add a fuzz test to ensure that #25 doesn't "invent" new Markdown events. However, this is proving more difficult than I thought since the underlying pulldown-cmark-to-cmark library isn't completely round-tripping the Markdown input. See https://github.com/Byron/pulldown-cmark-to-cmark/pull/55 for the discussion.

**Comment by mgeisler** [comment]
This was fixed with the 0.2.0 release! :rocket:  Please remember to run `mdbook-i18n-normalize` on our PO files to get the benefits of this.

**Comment by mgeisler** [comment]
@simonsan, thanks, you're very welcome! If you have a project which uses it, please add it to the README!
